src/concrete/ml/onnx/convert.py

```diff
@@ -235,13 +247,20 @@ def preprocess_onnx_model(onnx_model: onnx.ModelProto, check_model: bool) -> onn
             f"Available ONNX operators: {', '.join(sorted(IMPLEMENTED_ONNX_OPS))}"
         )
 
-    return equivalent_onnx_model
+    # Convert the first Gather node to a matrix multiplication with one-hot encoding
+    # In FHE, embedding is either a TLU or a matmul with a one-hot.
+    # The second case allows for leveled operation thus much faster.
+    onnx_preprocessing, equivalent_onnx_model = convert_first_gather_to_matmul(
```

Comment:
but we don't want that all the time though right ?

Reply:
Yes so I went for the solution where now we can have some part of the model done as pre-processing. The fact that we use an onnx makes it super easy to serialize and re-use on the client machine.

The other option is do the modification to the onnx in the `post_training.py` but since we manipulate onnx in convert.py I thought it was better do to it before.

Reply:
yes I think I get that part, but my point was more about that it looks like we are *always* converting any ONNX's first gather operator to matmul, since it's included in `preprocess_onnx_model` which is called everytime we convert to onnx, right ? I don't think that's intended !

Reply:
Ah sorry. Yes we always use that optimization. Also currently gather isn't supported with encrypted indices.

Reply:
ah ok, and is it somewhat related to https://github.com/zama-ai/concrete-ml-internal/issues/4413 then ?

Reply:
Yes good point! And also related to https://github.com/zama-ai/concrete-ml-internal/issues/4532 where we will have to make Gather nodes compatible in FHE one way or another.

Reply:
worth adding one or 2 fixme here then !

Reply:
Done thanks

Reply:
thanks